### PR TITLE
ADC pending command functionality

### PR DESCRIPTION
**Pull request by TeodoraMiu**
### Pull Request Overview

This pull request changes the way the ADC capsule receives commands from apps. Before this PR, the ADC capsule could only handle one command at a time. This PR would allow the ADC capsule to set an incoming command as pending if the requested ADC channel is already in use, and run the pending command after the currently active command finishes.

### Testing Strategy

This pull request was tested on a Microbit, using a fake ADC driver to simulate the case in which two commands can overlap.

### TODO or Help Wanted

Feedback is greatly appreciated.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: capsules/src/adc.rs [hunk 1]
```diff
@@ -653,16 +653,18 @@ impl<'a> AdcVirtualized<'a> {
...
-                .enter(processid, |app, _| {
-                    if self.current_app.is_none() {
-                        self.current_app.set(processid);
-                        let value = self.call_driver(command, channel);
-                        if value != Ok(()) {
-                            self.current_app.clear();
-                        }
-                        value
-                    } else {
+            if self.current_app.is_none() {
```
Comment (alexandruradovici) [inline capsules/src/adc.rs:656]: We renamed `appid` to `processid`, you might want to update your branch or just change the name of the variable here.

File: boards/components/src/adc.rs [hunk 2]
```diff
@@ -27,7 +27,6 @@ macro_rules! adc_component_static {
 macro_rules! adc_syscall_component_helper {
     ($($P:expr),+ $(,)?) => {{
         use kernel::count_expressions;
-        use kernel::static_init;
```
Comment (alexandruradovici) [inline boards/components/src/adc.rs:30]: Please keep this line, as the macro might be used in places where `static_init` is not defined.

File: capsules/src/adc.rs [hunk 3]
```diff
@@ -653,16 +653,18 @@ impl<'a> AdcVirtualized<'a> {
...
+                self.current_app.set(appid);
```
Comment (alexandruradovici) [inline capsules/src/adc.rs:657]:
```suggestion
                self.current_process.set(appid);
```
Please rename `current_app` to `current_process` in all places.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]
Please add a comment explaining how the syscall driver works and how applications can you it, describing the exact rules that apply.

**Review by phil-levis — APPROVED** [review]
Ah, nice. It looks like the machinery was there (`pending_command`) but it wasn't being used.

Can you explain your refactoring of `enqueue_command`?

**Review by alistair23 — APPROVED** [review]

**Review by alexandruradovici — APPROVED** [review]

**Comment by TeodoraMiu** [comment]
> Can you explain your refactoring of `enqueue_command`?

The enqueue_command function has been slightly changed: if the current process is none, there is no need to enter the grant. Therefore, I only enter the grant when there is an active process and the new command must be set as pending.

**Pull request merged**